plugin-server/src/worker/ingestion/event-pipeline/enrichExceptionEventStep.ts

```diff
@@ -0,0 +1,63 @@
+import { Counter } from 'prom-client'
+
+import { PreIngestionEvent } from '../../../types'
+import { EventPipelineRunner } from './runner'
+
+const EXTERNAL_FINGERPRINT_COUNTER = new Counter({
+    name: 'enrich_exception_events_external_fingerprint',
+    help: 'Counter for exceptions that already have a fingerprint',
+})
+
+const COULD_NOT_PARSE_STACK_TRACE_COUNTER = new Counter({
+    name: 'enrich_exception_events_could_not_parse_stack_trace',
+    help: 'Counter for exceptions where the stack trace could not be parsed',
+})
+
+const EXCEPTIONS_ENRICHED_COUNTER = new Counter({
+    name: 'enrich_exception_events_enriched',
+    help: 'Counter for exceptions that have been enriched',
+})
+
+export function enrichExceptionEventStep(
+    _runner: EventPipelineRunner,
+    event: PreIngestionEvent
+): Promise<PreIngestionEvent> {
+    if (event.event !== '$exception') {
+        return Promise.resolve(event)
+    }
+
+    const exceptionStack = event.properties['$exception_stack_trace_raw']
+    const fingerPrint = event.properties['$exception_fingerprint']
+    const type = event.properties['$exception_type']
+    const message = event.properties['$exception_message']
+    let firstFunction: string | null = null
+
+    if (fingerPrint) {
+        EXTERNAL_FINGERPRINT_COUNTER.inc()
+        return Promise.resolve(event)
+    }
+
+    try {
+        const parsedStack = JSON.parse(exceptionStack)
+        if (parsedStack.length > 0) {
+            firstFunction = parsedStack[0].function
+        }
+    } catch (e) {
+        COULD_NOT_PARSE_STACK_TRACE_COUNTER.inc()
+    }
+
+    const fingerprint = [
+        type ? 'TYPE' : false,
+        type,
+        message ? 'RAW_MESSAGE' : false,
+        message,
```

Comment:
For better grouping, would be cool if our API allows string interpolation such as:
logMessage("Error number {id}", {id: 500})
So we can group based on the string template, otherwise `Error number 500` and `Error number 400` wouldn't be grouped but they should because they are thrown because of an API error in the same call stack.
If we allow them to do just logMessage("Error number 500") without parameterized values, it's much harder for us to identify the templates.

Reply:
Agreed, that would be awesome! Another possibility would be wildcarding numbers so each of those messages comes through as `Error number *`. Has the potential to over group things so probably worth adding in a future version once this gets more complicated

Reply:
we could use a glob-like syntax but that will hide the original values, the parametrization allows us better grouping but also shows the original titles when navigating through the issues.